Framework/Core/include/Framework/MessageContext.h

```diff
@@ -387,24 +397,42 @@ class MessageContext
 
   /// Schedule a context object for sending.
   /// The object is considered complete at this point and is sent directly through the dispatcher callback
-  /// of the context if initialized. It is added to the list of messages otherwise.
+  /// of the context if initialized.
   void schedule(Messages::value_type&& message)
   {
-    if (mDispatchCallback) {
-      mDispatchCallback(std::move(message->finalize()), message->channel(), DefaultChannelIndex);
-      return;
+    auto const* header = message->header();
+    if (header == nullptr) {
+      throw std::logic_error("No valid header message found");
+    }
+    mScheduledMessages.emplace_back(std::move(message));
+    if (mDispatchControl.dispatch != nullptr) {
+      // send all scheduled messages if there is no trigger callback or its result is true
+      if (mDispatchControl.trigger == nullptr || mDispatchControl.trigger(*header)) {
+        std::unordered_map<std::string, FairMQParts> outputs;
```

Comment:
Can we do without the string allocation? This will show up in the benchmarks which use high data rates (e.g. the QC one).

Reply:
correct, I will implement this based in indices. Do we have a dedicated benchmark test? Going to add one for the sending of messages otherwise.